File: mobile/src/main/java/com/windscribe/mobile/custom_view/CustomDrawableCrossFadeFactory.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.custom_view;


import android.graphics.drawable.Drawable;

import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.request.transition.DrawableCrossFadeTransition;
import com.bumptech.glide.request.transition.Transition;
import com.bumptech.glide.request.transition.TransitionFactory;

@SuppressWarnings("WeakerAccess")
public class CustomDrawableCrossFadeFactory implements TransitionFactory<Drawable> {

    /**
     * A Builder for {@link CustomDrawableCrossFadeFactory}.
     */
    public static class Builder {

        private final int durationMillis;

        private boolean isCrossFadeEnabled;

        public Builder(int durationMillis) {
            this.durationMillis = durationMillis;
        }

        public CustomDrawableCrossFadeFactory build() {
            return new CustomDrawableCrossFadeFactory(durationMillis, isCrossFadeEnabled);
        }

        public Builder setCrossFadeEnabled(boolean isCrossFadeEnabled) {
            this.isCrossFadeEnabled = isCrossFadeEnabled;
            return this;
        }
    }

    private final int duration;

    private final boolean isCrossFadeEnabled;

    private DrawableCrossFadeTransition resourceTransition;

    protected CustomDrawableCrossFadeFactory(int duration, boolean isCrossFadeEnabled) {
        this.duration = duration;
        this.isCrossFadeEnabled = isCrossFadeEnabled;
    }

    @Override
    public Transition<Drawable> build(DataSource dataSource, boolean isFirstResource) {
        return getResourceTransition();
    }

    private Transition<Drawable> getResourceTransition() {
        if (resourceTransition == null) {
            resourceTransition = new DrawableCrossFadeTransition(duration, isCrossFadeEnabled);
        }
        return resourceTransition;
    }
}

File: mobile/src/main/java/com/windscribe/mobile/custom_view/CustomDrawableCrossFadeFactory.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.custom_view

import android.graphics.drawable.Drawable
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.request.transition.DrawableCrossFadeTransition
import com.bumptech.glide.request.transition.Transition
import com.bumptech.glide.request.transition.TransitionFactory

open class CustomDrawableCrossFadeFactory protected constructor(
    private val duration: Int, private val isCrossFadeEnabled: Boolean
) : TransitionFactory<Drawable> {
    /**
     * A Builder for [CustomDrawableCrossFadeFactory].
     */
    class Builder(private val durationMillis: Int) {
        private var isCrossFadeEnabled = false
        fun build(): CustomDrawableCrossFadeFactory {
            return CustomDrawableCrossFadeFactory(durationMillis, isCrossFadeEnabled)
        }

        fun setCrossFadeEnabled(isCrossFadeEnabled: Boolean): Builder {
            this.isCrossFadeEnabled = isCrossFadeEnabled
            return this
        }
    }

    private var resourceTransition: DrawableCrossFadeTransition? = null
        get() {
            if (field == null) {
                field = DrawableCrossFadeTransition(duration, isCrossFadeEnabled)
            }
            return field
        }

    override fun build(dataSource: DataSource, isFirstResource: Boolean): Transition<Drawable>? {
        return resourceTransition
    }
}